olp-cpp-sdk-dataservice-read/src/TaskContext.h

```diff
@@ -0,0 +1,181 @@
+#include "olp/core/client/CancellationToken.h"
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+class TaskContext {
+ public:
+  template <typename Exec, typename Callback>
+  static TaskContext create(Exec execute_func, Callback callback) {
+    TaskContext context;
+    context.SetExecutors(std::move(execute_func), std::move(callback));
+    return context;
+  }
+
+  void Execute() const { impl_->Execute(); }
+
+  void BlockingCancel(
+      std::chrono::milliseconds timeout = std::chrono::seconds(60)) const {
+    impl_->BlockingCancel(timeout);
+  }
+
+  client::CancellationToken CancelToken() const { return impl_->CancelToken(); }
+
+  bool operator==(const TaskContext& other) const {
+    return impl_ == other.impl_;
+  }
+
+ protected:
+  friend struct TaskContextHash;
+
+  TaskContext() = default;
+
+  template <typename Exec, typename Callback,
+            typename ExecResult = typename std::result_of<
+                Exec(client::CancellationContext)>::type>
+  void SetExecutors(Exec execute_func, Callback callback) {
+    impl_ = std::make_shared<TaskContextImpl<typename ExecResult::Result>>(
+        std::move(execute_func), std::move(callback));
+  }
+
+  class Impl {
+   public:
+    virtual ~Impl() = default;
+    virtual void Execute() = 0;
+    virtual void BlockingCancel(std::chrono::milliseconds timeout) = 0;
+    virtual client::CancellationToken CancelToken() = 0;
+  };
+
+  template <typename T>
+  class TaskContextImpl : public Impl {
+   public:
+    using Response = client::ApiResponse<T, client::ApiError>;
+    using ExecuteFunc = std::function<Response(client::CancellationContext)>;
+    using UserCallback = std::function<void(Response)>;
+
+    TaskContextImpl(ExecuteFunc execute_func, UserCallback callback)
+        : execute_func_(std::move(execute_func)),
+          callback_(std::move(callback)),
+          state_{State::PENDING} {}
+
+    ~TaskContextImpl() {}
+
+    void Execute() override {
+      State expected_state = State::PENDING;
+
+      if (!state_.compare_exchange_strong(expected_state, State::IN_PROGRESS)) {
+        return;
+      }
+
+      // Moving the user callback and function guarranty that they will be
+      // executed excactly 1 time
+      ExecuteFunc function = nullptr;
+      UserCallback callback = nullptr;
+
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        function = std::move(execute_func_);
+        callback = std::move(callback_);
+      }
+
+      Response user_response =
+          client::ApiError(client::ErrorCode::Cancelled, "Cancelled");
+
+      if (function && !context_.IsCancelled()) {
+        auto response = function(context_);
+        // Cancel could occur during function execution, in that case we ignore
+        // the response.
+        if (!context_.IsCancelled()) {
+          user_response = std::move(response);
+        }
+      }
+
+      if (callback) {
+        callback(std::move(user_response));
```

Comment:
What if user calls from other thread the `BlockingCancel` right before the callback execution? 

Should we handle that case as well?